website/docs/externaldata.md

```diff
@@ -0,0 +1,164 @@
+---
+id: externaldata
+title: External Data
+---
+
+> This feature is still in alpha stage, so the final form can still change (feedback is welcome!). Mutation is not yet supported with external data.
+
+## Motivation
+
+Gatekeeper provides various means to mutate and validate Kubernetes resources. However, in many of these scenarios this data is either built-in, static or user defined. With external data feature, we are enabling Gatekeeper to interface with various external data sources, such as image registries, using a provider-based model.
+
+A similar way to connect with an external data source can be done today using OPA's built-in `http.send` functionality. However, there are limitations to this approach.
+- Gatekeeper does not support Rego policies for mutation, which cannot use the OPA `http.send` built-in function.
+- Security concerns due to:
+  - if template authors are not trusted, it will potentially give template authors access to the in-cluster network.
+  - if template authors are trusted, authors will need to be careful on how rego is written to avoid injection attacks.
+
+Using the external data solution provides benefits, such as:
+- Addresses security concerns by:
+  - Restricting which hosts a user can access.
+  - Providing an interface for making requests, which allows Gatekeeper to better handle things like escaping strings.
+- Addresses common patterns with a single provider, such as image SHA mutation, that potential scenarios (such as validating image signatures or vulnerabilities) may utilize.
+- Provider model creates a common interface for extending Gatekeeper with external data.
+  - Developers and consumers of data sources can rely on that common protocol to ease authoring of both constraint templates and data sources.
+  - Makes change management easier as users of an external data provider should be able to tell whether upgrading it will break existing constraint templates. (once external data API is stable, our goal is to have that answer always be "no")
+- Performance benefits as Gatekeeper can now directly control caching and which values are significant for caching, which increases the likelihood of cache hits.
+  - For mutation, we can batch requests via lazy evaluation.
+  - For validation, we make batching easier via function design.
```

Comment:
> For validation, we make batching easier via function design

what does "function design" mean here?

Reply:
I think this means `external_data` rego function receives a list. @maxsmythe is this correct? should we clarify?

Reply:
yep, the function signature makes it easier to batch calls in Rego by accepting a list.

Reply:
added clarification that it's for rego func